9.4/root/usr/share/container-scripts/postgresql/common.sh

```diff
@@ -218,3 +223,165 @@ function wait_for_postgresql_master() {
     sleep 1
   done
 }
+
+
+run_pgupgrade ()
+(
+  optimized=false
+  old_raw_version=${POSTGRESQL_PREV_VERSION//\./}
+  new_raw_version=${POSTGRESQL_VERSION//\./}
+
+  if test "$old_raw_version" = 92; then
+    old_collection=postgresql92
+  else
+    old_collection=rh-postgresql$old_raw_version
+  fi
+
+  old_pgengine=/opt/rh/$old_collection/root/usr/bin
+  new_pgengine=/opt/rh/rh-postgresql${new_raw_version}/root/usr/bin
+  PGDATA_new="${PGDATA}-new"
+
+  printf >&2 "\n==========  \$PGDATA upgrade: %s -> %s  ==========\n\n" \
+             "$POSTGRESQL_PREV_VERSION" \
+             "$POSTGRESQL_VERSION"
+
+  info_msg () { printf >&2 "\n===>  $*\n\n" ;}
+
+  # pg_upgrade writes logs to cwd, so go to the persistent storage first
+  cd "$HOME"/data
+
+  # disable this because of scl_source, 'set +u' just makes the code ugly
+  # anyways
+  set +u
+
+  # we need to have the old SCL enabled, otherwise the $old_pgengine is not
+  # working.  The scl_source script doesn't pay attention to non-zero exit
+  # statuses, so use 'set +e'.
+  set +e
+  source scl_source enable $old_collection
+  set -e
+
+  case $POSTGRESQL_UPGRADE in
+    copy) # we accept this
+      ;;
+    hardlink)
+      optimized=:
+      ;;
+    *)
+      echo >&2 "Unsupported value: \$POSTGRESQL_UPGRADE=$POSTGRESQL_UPGRADE"
+      false
+      ;;
+  esac
+
+  # Ensure $PGDATA_new doesn't exist yet, so we can immediately remove it if
+  # there's some problem.
+  test ! -e "$PGDATA_new"
+
+  # initialize the database
+  info_msg "Initialize new data directory; we will migrate to that."
+  initdb_cmd=( initdb_wrapper "$new_pgengine"/initdb "$PGDATA_new" )
+  eval "\${initdb_cmd[@]} ${POSTGRESQL_UPGRADE_INITDB_OPTIONS-}" || \
+    { rm -rf "$PGDATA_new" ; false ; }
+
+  upgrade_cmd=(
+      "$new_pgengine"/pg_upgrade
+      "--old-bindir=$old_pgengine"
+      "--new-bindir=$new_pgengine"
+      "--old-datadir=$PGDATA"
+      "--new-datadir=$PGDATA_new"
+  )
+
+  # Dangerous --link option, we loose $DATADIR if something goes wrong.
+  ! $optimized || upgrade_cmd+=(--link)
+
+  # User-specififed options for pg_upgrade.
+  eval "upgrade_cmd+=(${POSTGRESQL_UPGRADE_PGUPGRADE_OPTIONS-})"
+
+  # the upgrade
+  info_msg "Starting the pg_upgrade process."
+
+  # Once we stop support for PostgreSQL 9.4, we don't need
+  # REDHAT_PGUPGRADE_FROM_RHEL hack as we don't upgrade from 9.2 -- that means
+  # that we don't need to fiddle with unix_socket_director{y,ies} option.
+  REDHAT_PGUPGRADE_FROM_RHEL=1 \
+  "${upgrade_cmd[@]}" || { rm -rf "$PGDATA_new" && false ; }
+
+  # Move the important configuration and remove old data.  This is highly
+  # careless, but we can't do more for this over-automatized process.
```

Comment:
I've also had problems with this.

The real issue comes when a new version breaks compatibility with parameters, this happened from 9.4 to 9.5  (specifically for the checkpoint_segments config variable moving to max/min_wal_size). In this case if a user specified those settings in their conf the server would not start after upgrade.

I think this (replacing the config) is still the most simple solution for now, but as a follow-up/enhancement could we utilize the concept of an [include directory](https://www.postgresql.org/docs/9.4/static/config-setting.html#CONFIG-INCLUDES) for postgresql configs?

This would separate user changes from the default config shipped with a particular version and might mitigate the issues encountered here. Additionally this would map nicely to using OpenShift/Kubernetes [ConfigMaps](https://docs.openshift.org/latest/dev_guide/configmaps.html) for user PG config overrides.

Reply:
@carbonin wrote:
> I think this (replacing the config) is still the most simple solution for now, but as a follow-up/enhancement could we utilize the concept of an include directory for postgresql configs?
Try:
```
cat data/userdata/postgresql.conf | grep ^include
include '/var/lib/pgsql/openshift-custom-postgresql.conf'
```

So in postgresql.conf should only be the default options + the include statement.

Reply:
But yeah, I really think that the `pg_upgrade` won't work well for everybody;  and that's why we make the documentation saying that users should make themselves comfortable with manual rollback in case of any failure.